src/app/qgsmaptooldigitizefeature.cpp

```diff
@@ -138,21 +138,32 @@ void QgsMapToolDigitizeFeature::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
       return;
     }
 
-    QgsPointXY savePoint; //point in layer coordinates
+    QgsPoint savePoint; //point in layer coordinates
+    bool isMatchPointZ = false;
     try
     {
       QgsPoint fetchPoint;
       int res;
       res = fetchLayerPoint( e->mapPointMatch(), fetchPoint );
+      if ( QgsWkbTypes::hasZ( fetchPoint.wkbType() ) )
+        isMatchPointZ = true;
+
       if ( res == 0 )
       {
-        savePoint = QgsPointXY( fetchPoint.x(), fetchPoint.y() );
+        if ( isMatchPointZ )
+          savePoint = QgsPoint( QgsWkbTypes::PointZ, fetchPoint.x(), fetchPoint.y(), fetchPoint.z() );
+        else
+          savePoint = QgsPoint( fetchPoint.x(), fetchPoint.y() );
       }
       else
       {
-        savePoint = toLayerCoordinates( vlayer, e->mapPoint() );
+        QgsPointXY layerPoint = toLayerCoordinates( vlayer, e->mapPoint() );
+        if ( isMatchPointZ )
+          savePoint = QgsPoint( QgsWkbTypes::PointZ, layerPoint.x(), layerPoint.y(), fetchPoint.z() );
+        else
+          savePoint = QgsPoint( layerPoint.x(), layerPoint.y() );
       }
-      QgsDebugMsg( "savePoint = " + savePoint.toString() );
+      QgsDebugMsg( "savePoint = " + savePoint.asWkt() );
```

Comment:
Do we need that?

Reply:
I don't think so, even if it was useful to me during debugging.

Reply:
removed